### PR TITLE
HTML Lorem

### PR DESCRIPTION
**Pull request by rudkjobing**
A provider for faking HTML that should be easy to expand and improve.

Let me know if I have overlooked anything.

Regards Steffen.

### CODE REVIEW

**Comment by fzaninotto** [comment]
Interesting! One question: why do you use custom classes rather than PHP DODDocument API?

**Comment by mpdude** [comment]
Does this aim at creating "complete" HTML documents, or can we also get just a bunch of paragraphs (`<p>`), maybe with some mixed-in `<b>` and `<i>`?

**Comment by rudkjobing** [comment]
@fzaninotto Great question, the answer is ignorance :-). I have just committed a version that only uses DomDocument.
@mpdude I wouldn't mind if that were to be added.

**Comment by fzaninotto** [comment]
Great! Could you paste a sample result in the PR description? Also, Can you add a line about the new formatter in the README?

**Comment by rudkjobing** [comment]
Cool. Think I have done what you asked. If not, let me know.

File: readme.md [hunk 1]
```diff
@@ -298,6 +299,11 @@ Each of the generator properties (like `name`, `address`, and `lorem`) are calle
...
...em aliquid saepe laudantium facere eos sunt dolor.</td><td>Est eos quis laboriosam officia expedita repellendus quia natus.</td><td>Et neque delectus quod fugit enim repudiandae qui.</td><td>Fugit soluta sit facilis facere repellat culpa magni voluptatem maiores tempora.</td></tr><tr><td>Enim dolores doloremque.</td><td>Assumenda voluptatem eum perferendis exercitationem.</td><td>Quasi in fugit deserunt ea perferendis sunt nemo consequatur dolorum soluta.</td><td>Maxime repellat qui numquam voluptatem est modi.</td><td>Alias rerum rerum hic hic eveniet.</td></tr><tr><td>Tempore voluptatem.</td><td>Eaque.</td><td>Et sit quas fugit iusto.</td><td>Nemo nihil rerum dignissimos et esse.</td><td>Repudiandae ipsum numquam.</td></tr><tr><td>Nemo sunt quia.</td><td>Sint tempore est neque ducimus harum sed.</td><td>Dicta placeat atque libero nihil.</td><td>Et qui aperiam temporibus facilis eum.</td><td>Ut dolores qui enim et maiores nesciunt.</td></tr><tr><td>Dolorum totam sint debitis saepe laborum.</td><td>Quidem corrupti ea.</td><td>Cum voluptas quod.</td><td>Possimus consequatur quasi dolorem ut et.</td><td>Et velit non hic labore repudiandae quis.</td></tr></tbody></table></body></html>
```
Comment (fzaninotto) [inline readme.md]: Then I don't understand what "depth" and "width" mean...

File: test/Faker/Provider/HtmlLoremTest.php [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+namespace Faker\Test\Provider;
+
+use Faker\Generator;
+use Faker\Provider\HtmlLorem;
+
+class HtmlLoremTest extends \PHPUnit_Framework_TestCase
+{
+
+    public function testProvider()
```
Comment (fzaninotto) [inline test/Faker/Provider/HtmlLoremTest.php]: `testRandomHtmlReturnsValidHTMLString()`, and use DOM validation to prove it.

File: src/Faker/Provider/HtmlLorem.php [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+                break;
+        }
+    }
+
+    private static function addRandomAttribute(\DOMElement $node)
+    {
+        $rand = mt_rand(1, 2);
+        switch ($rand) {
+            case 1:
+                $node->setAttribute("class", Lorem::word());
```
Comment (fzaninotto) [inline src/Faker/Provider/HtmlLorem.php]: You shouldn't use `Lorem` statically, but use `$this->generator->word` instead.

File: src/Faker/Provider/HtmlLorem.php [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+        $node = $element->ownerDocument->createElement(HtmlLorem::A_TAG);
+        $node->setAttribute("href", Internet::safeEmailDomain());
+        $node->appendChild($text);
+        $element->appendChild($node);
+    }
+
+    private static function addRandomTitle(\DOMElement $element, $maxLength = 10)
+    {
+        $text = $element->ownerDocument->createTextNode(Lorem::sentence(mt_rand(1, $maxLength)));
+        $node = $element->ownerDocument->createElement(HtmlLorem::TITLE_TAG);
```
Comment (fzaninotto) [inline src/Faker/Provider/HtmlLorem.php]: use `static` there instead and for all constants

File: readme.md (hunk 1 above)
Comment (rudkjobing) [inline readme.md]: I just used them to limit the size of the document. Depth is how deep the tree can get in total. Width is how deep each layer can get.

File: test/Faker/Provider/HtmlLoremTest.php (hunk 2 above)
Comment (rudkjobing) [inline test/Faker/Provider/HtmlLoremTest.php]: Done

File: src/Faker/Provider/HtmlLorem.php (hunk 3 above)
Comment (rudkjobing) [inline src/Faker/Provider/HtmlLorem.php]: how would i use $this in a static method? The mix of static / non-static in the faker is slightly confusing.

File: readme.md (hunk 1 above)
Comment (fzaninotto) [inline readme.md]: You should explain it in the README

File: src/Faker/Provider/HtmlLorem.php (hunk 3 above)
Comment (fzaninotto) [inline src/Faker/Provider/HtmlLorem.php]: make it non-static

File: src/Faker/Provider/HtmlLorem.php (hunk 4 above)
Comment (fzaninotto) [inline src/Faker/Provider/HtmlLorem.php]: I don't, but some developers do.

File: test/Faker/Provider/HtmlLoremTest.php [hunk 5]
```diff
@@ -17,4 +17,14 @@ public function testProvider()
...
     }
 
+    public function testRandomHtmlReturnsValidHTMLString(){
+        $faker = new Generator();
+        $faker->addProvider(new HtmlLorem($faker));
+        $node = $faker->randomHtml(6, 10);
+        $dom = new \DOMDocument();
+        $error = $dom->loadHTML($node);
+        $this->assertTrue($error);
+
```
Comment (fzaninotto) [inline test/Faker/Provider/HtmlLoremTest.php]: Extra line feed

File: src/Faker/Provider/HtmlLorem.php (hunk 3 above)
Comment (rudkjobing) [inline src/Faker/Provider/HtmlLorem.php]: Done

File: src/Faker/Provider/HtmlLorem.php (hunk 4 above)
Comment (rudkjobing) [inline src/Faker/Provider/HtmlLorem.php]: Done

File: test/Faker/Provider/HtmlLoremTest.php (hunk 5 above)
Comment (rudkjobing) [inline test/Faker/Provider/HtmlLoremTest.php]: Removed

File: readme.md (hunk 1 above)
Comment (rudkjobing) [inline readme.md]: I have attempted to make it a bit more clear.

**Comment by fzaninotto** [comment]
Awesome, thanks a lot for this one!

**Comment by bwach** [comment]
great job, thanks @rudkjobing, @fzaninotto

**Comment by rudkjobing** [comment]
@fzaninotto my pleasure! @bwach thanks! :)

**Comment by mpdude** [comment]
Can we get a 1.7 release with this please?

**Comment by ceesvanegmond** [comment]
👍

**Comment by graciano** [comment]
I tried to use what I think that came from this PR documented [here](https://github.com/fzaninotto/Faker#fakerproviderhtmllorem) in the readme and it didn't work.
Since the last release was before this PR, I think it's just a matter of version (I did a require `dev-master` in composer and it worked), so I'm posting this warning.

**Comment by mpdude** [comment]
@rudkjobing do you think you could extend this to alternatively produce just fragments, that is, the markup you'd usually expect within the body tag?

**Comment by rudkjobing** [comment]
@mpdude Yes absolutely. Anything else you would like?

**Comment by mpdude** [comment]
If it were 8 hours later, I'd ask for a beer. But...

Honestly, thanks for your contribution!

**Comment by rudkjobing** [comment]
I can supply that as well, provided you are in Aarhus/DK ;)

**Comment by mpdude** [comment]
@rudkjobing hopefully I will remember that and get back to you should I happen to visit Midtjylland sometime soon :-) 🍻

**Pull request merged**